File: python3/238.py
class Solution:
    def productExceptSelf(self, n: List[int]) -> List[int]:
        o = -1
        p = 1
        for i in range(len(n)):
            if n[i] == 0:
                if o != -1:
                    return [0] * len(n)
                o = i
            else:
                p *= n[i]
        if o != -1:
            r = [0] * len(n)
            r[o] = p
            return r
        return [int(p / n[i]) for i in range(len(n))]

File: cpp/238.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& n) {
        int l = n.size(), o = -1, p = 1;
        vector<int> r(l);
        for (int i = 0; i < l; i++)
            if (n[i] == 0) {
                if (o != -1)
                    return r;
                o = i;
            } else
                p *= n[i];
        if (o != -1) {
            r[o] = p;
            return r;
        }
        for (int i = 0; i < l; i++)
            r[i] = p / n[i];
        return r;
    }
};
